packages/amplify-graphql-api-construct/src/types.ts

```diff
@@ -0,0 +1,222 @@
+
+export type IAMAuthorizationConfig = {
+  identityPoolId?: string;
+  authRole?: IRole;
+  unauthRole?: IRole;
+  adminRoles?: IRole[];
+};
+
+export type UserPoolAuthorizationConfig = {
+  userPool: IUserPool;
+};
+
+export type OIDCAuthorizationConfig = {
+  oidcProviderName: string;
+  oidcIssuerUrl: string;
+  clientId?: string;
+  tokenExpiryFromAuth: Duration;
+  tokenExpiryFromIssue: Duration;
+};
+
+export type ApiKeyAuthorizationConfig = {
+  description?: string;
+  expires: Duration;
+};
+
+export type LambdaAuthorizationConfig = {
+  function: IFunction;
+  ttl: Duration;
+};
+
+/**
+ * Authorization Config to apply to the API.
+ * At least one config must be provided, and if more than one are provided,
+ * a defaultAuthMode must be specified.
+ * For more information on Amplify API auth, refer to https://docs.amplify.aws/cli/graphql/authorization-rules/#authorization-strategies
+ */
+export type AuthorizationConfig = {
+  /**
+   * Default auth mode to provide to the API, required if more than one config type is specified.
+   */
+  defaultAuthMode?:
+    | 'AWS_IAM'
+    | 'AMAZON_COGNITO_USER_POOLS'
+    | 'OPENID_CONNECT'
+    | 'API_KEY'
+    | 'AWS_LAMBDA';
+
+  /**
+   * IAM Auth config, required if an 'iam' auth provider is specified in the API.
+   * Applies to 'public' and 'private' auth strategies.
+   */
+  iamConfig?: IAMAuthorizationConfig;
+
+  /**
+   * Cognito UserPool config, required if a 'userPools' auth provider is specified in the API.
+   * Applies to 'owner', 'private', and 'group' auth strategies.
+   */
+  userPoolConfig?: UserPoolAuthorizationConfig;
+
+  /**
+   * Cognito OIDC config, required if a 'oidc' auth provider is specified in the API.
+   * Applies to 'owner', 'private', and 'group' auth strategies.
+   */
+  oidcConfig?: OIDCAuthorizationConfig;
+
+  /**
+   * AppSync API Key config, required if a 'apiKey' auth provider is specified in the API.
+   * Applies to 'public' auth strategy.
+   */
+  apiKeyConfig?: ApiKeyAuthorizationConfig;
+
+  /**
+   * Lambda config, required if a 'function' auth provider is specified in the API.
+   * Applies to 'custom' auth strategy.
+   */
+  lambdaConfig?: LambdaAuthorizationConfig;
+};
+
+/**
+ * Schema representation for transformation. Accepts either a raw string, single, or array of appsync SchemaFile objects.
+ */
+export type AmplifyGraphQlApiSchema =
+  | SchemaFile
+  | SchemaFile[]
+  | string;
+
+/**
+ * Input props for the AmplifyGraphQLApi construct. Specifies what the input to transform into an API, and configurations for
+ * the transformation process.
+ */
+export type AmplifyGraphQlApiProps = {
+  /**
+   * The schema to transform in a full API.
+   */
+  schema: AmplifyGraphQlApiSchema;
+
+  /**
+   * Override value if an env is provided for distinguishing between transformed apis.
+   * Default: NONE
+   */
+  envOverride?: string;
+
+  /**
+   * Name to be used for the appsync api.
+   * Default: construct id.
```

Comment:
I remember for other CDK constructs (like s3) if I do not pass a name attribute but only initiate it with a construct id, the final name of it will be the construct id appending a uuid to make it globally unique. If the appsync api name is only unique within the account, it should be fine to use the original construct id

Reply:
It is unique within the account, but we should discuss a bit more on whether we need/want to suffix here to be idiomatic. This is I think related to the env discussion below. I'll spin that out into a separate spike/task.